src/main/java/com/snowflake/kafka/connector/internal/StageFilesProcessor.java

```diff
@@ -0,0 +1,505 @@
+    private final KCLogger LOGGER = new KCLogger(StageFilesProcessor.class.getName());
+    private final String pipeName;
+    private final String tableName;
+    private final String stageName;
+    private final String prefix;
+    private final SnowflakeConnectionService conn;
+    private final AtomicReference<Thread> runnerThread = new AtomicReference<>();
+    private final TimeSupplier currentTimeSupplier;
+    private final SnowflakeIngestionService ingestionService;
+    private final SnowflakeTelemetryPipeStatus pipeTelemetry;
+    private final SnowflakeTelemetryService telemetryService;
+    private final ExecutorService cleanerExecutor;
+    private final FilteringPredicates filters;
+
+    /**
+     * Client interface for the StageFileProcessor - allows thread safe registration of new files and notifies
+     * the processor about offset update.
+     */
+    interface ProgressRegister {
+        void registerNewStageFile(String fileName);
+        void newOffset(long offset);
+        SnowflakeTelemetryPipeCreation getPipeCreation();
+        SnowflakeTelemetryPipeStatus getPipeTelemetry();
+        SnowflakeTelemetryService getTelemetry();
+        void close();
+    }
+
+    /**
+     * This interface is a wrapper on System.currentTimeMillis() - required for testing purposes
+     * */
+    @FunctionalInterface
+    interface TimeSupplier {
+        long currentTime();
+    }
+
+    /**
+     * This interface is a wrapper on Thread::sleep() - required for testing purposes
+     */
+    @FunctionalInterface
+    interface TaskAwaiter {
+        void await(long duration) throws InterruptedException;
+    }
+
+    public StageFilesProcessor(
+            String pipeName,
+            String tableName,
+            String stageName,
+            String prefix,
+            SnowflakeConnectionService conn,
+            SnowflakeIngestionService ingestionService,
+            SnowflakeTelemetryPipeStatus pipeTelemetry,
+            SnowflakeTelemetryService telemetryService)  {
+        this(pipeName, tableName, stageName, prefix, conn, ingestionService, pipeTelemetry, telemetryService, System::currentTimeMillis);
+    }
+
+    @VisibleForTesting
+    StageFilesProcessor(
+            String pipeName,
+            String tableName,
+            String stageName,
+            String prefix,
+            SnowflakeConnectionService conn,
+            SnowflakeIngestionService ingestionService,
+            SnowflakeTelemetryPipeStatus pipeTelemetry,
+            SnowflakeTelemetryService telemetryService,
+            TimeSupplier currentTimeSupplier) {
+        this.pipeName = pipeName;
+        this.tableName = tableName;
+        this.stageName = stageName;
+        this.prefix = prefix;
+        this.conn = conn;
+        this.currentTimeSupplier = currentTimeSupplier;
+        this.ingestionService = ingestionService;
+        this.telemetryService = telemetryService;
+        this.cleanerExecutor = ForkJoinPool.commonPool();
```

Comment:
Do we really want to use a common `ForkJoinPool`? I assume it can be used by multiple different connectors (not only Kafka ones) if they're running in the same kafka connect instance.

Reply:
and? i'd rather use the threads which are already there, present and ready for work, rather than spawn new ones. 
the moment thread from the pool is stuck on the I/O or timed wait state and pool is running out of threads - it will automatically add a new one. 
also - the lifetime of the sink service is controlled by kafka - it may go up and down quite frequently - in this case, not managing our own threads is even better idea.